File: day-12/python/main.py
import sys
import numpy as np
import math


def parse_args():
    if len(sys.argv) != 2:
        sys.exit('I need a file')
    return sys.argv[1]


def parse_input(filename):
    grid = []
    start = None
    end = None
    i = 0
    with open(sys.argv[1]) as fd:
        for line in fd:
            j = 0
            x = None
            row = []
            for c in line.strip():
                if c == 'S':
                    start = (i, j)
                    x = 0
                elif c == 'E':
                    end = (i, j)
                    x = ord('z') - ord('a')
                else:
                    x = ord(c) - ord('a')
                j += 1
                row.append(x)
            grid.append(row)
            i += 1
    assert start and end
    return grid, start, end


def idx(i, j, cols):
    return i * cols + j


def to_graph(grid, rows, cols):
    cmp = lambda other, cur: (other - cur) < 2
    g = np.zeros((rows * cols, rows * cols), dtype=int)
    for i in range(rows):
        for j in range(cols):
            v = grid[i][j]
            if i != 0 and cmp(grid[i - 1][j], v):
                g[idx(i, j, cols), idx(i - 1, j, cols)] = 1
            if i != (rows - 1) and cmp(grid[i + 1][j], v):
                g[idx(i, j, cols), idx(i + 1, j, cols)] = 1
            if j != 0 and cmp(grid[i][j - 1], v):
                g[idx(i, j, cols), idx(i, j - 1, cols)] = 1
            if j != (cols - 1) and cmp(grid[i][j + 1], v):
                g[idx(i, j, cols), idx(i, j + 1, cols)] = 1
    return g


def unvisited_with_min_distance(unvisited, dists):
    mind, mini = None, None
    i = 0
    for t in unvisited:
        if i == 0:
            mind = dists[t]
            mini = t
            i += 1
        else:
            if dists[t] < mind:
                mind = dists[t]
                mini = t
    return mini


def dijkstra(g, start):
    n = g.shape[0]
    # keep track of the unvisited nodes
    unvisited = set(range(n))
    # store distances in d
    d = np.zeros((n,), dtype=int) + int(1e8)
    d[start] = 0
    for i in range(n):
        u = unvisited_with_min_distance(unvisited, d)
        unvisited.remove(u)
        for v in range(n):
            if g[u, v] > 0 and (v in unvisited):
                d[v] = min(d[v], d[u] + g[u, v])
    return d


def problem_1(g, start, end):
    dists = dijkstra(g, start)
    return dists[end]


def problem_2(g, grid, end, cols):
    dists = dijkstra(np.transpose(g), end)
    return min([dists[idx(i, j, cols)] for ((i, j), el) in np.ndenumerate(grid) \
                if el == 0])


def main():
    # parse data
    filename = parse_args()
    grid, start, end = parse_input(filename)
    rows, columns = len(grid), len(grid[0])

    # create the graph
    g = to_graph(grid, rows, columns)
    start, end = idx(*start, columns), idx(*end, columns)

    print("Day 12, problem 1: ", problem_1(g, start, end))
    print("Day 12, problem 2: ", problem_2(g, grid, end, columns))


if __name__ == '__main__':
    main()

File: day-12/rust/src/main.rs
use std::env;
use std::fs::File;
use std::cmp;
use std::io::{self, BufRead};
use std::process;
use petgraph::graph::{NodeIndex, Graph};
use petgraph::algo::dijkstra;

fn parse_args() -> String {
    let mut args = env::args();
    if args.len() != 2 {
        eprintln!("Invalid arguments: I need the input file and nothing else");
        process::exit(1);
    }
    args.nth(1).expect("Already checked")
}

fn create_reader(filename: &String) -> impl Iterator<Item = String> {
    let file = File::open(&filename).unwrap_or_else(|_| {
        eprintln!("Cannot open file: {}", &filename);
        process::exit(1);
    });
    io::BufReader::new(file)
        .lines()
        .map(|l| l.unwrap())
}

struct Node {
    i: usize,
    j: usize
}

fn to_code(c: char) -> u32 {
    c as u32 - 'a' as u32
}

fn check_columns(grid: &Vec<Vec<u32>>) -> usize {
    assert!(!grid.is_empty());
    let columns = grid[0].len();
    for i in 1..grid.len() {
        if columns != grid[i].len() {
            panic!("Invalid columns in line: {}", i + 1);
        }
    }
    columns
}

fn parse_input(filename: &String) -> (Vec<Vec<u32>>, Node, Node, usize) {
    let lines = create_reader(filename);
    let mut start: Option<Node> =  None;
    let mut end: Option<Node> = None;
    let mut grid = Vec::<Vec<u32>>::new();
    let mut i = 0;
    for line in lines {
        grid.push(Vec::new());
        for (j, c) in line.trim().chars().enumerate() {
            let v: u32 = match c {
                'S' => {
                    start = Some(Node{ i: i, j: j });
                    to_code('a')
                },
                'E' => {
                    end = Some(Node{ i: i, j: j });
                    to_code('z')
                },
                _ => to_code(c)
            };
            grid[i].push(v);
        }
        i += 1;
    }
    let columns = check_columns(&grid);
    if !start.is_some() || !end.is_some() {
        eprintln!("Invalid input. Start of finish notes are not annotated");
        process::exit(1);
    }
    (grid, start.unwrap(), end.unwrap(), columns)
}

fn to_idx(i: usize, j: usize, cols: usize) -> usize {
    i * cols + j
}

/// Converts the given grid to a graph based on the transition rules for hill
/// climbing
fn to_graph(grid: &Vec<Vec<u32>>, cols: usize) -> Graph<u32, u32> {
    let mut g = Graph::<u32, u32>::new();
    for i in 0..grid.len() {
        for j in 0..grid[0].len() {
            let idx = to_idx(i, j, cols) as u32;
            let v = grid[i][j];
            let mut new_edges = Vec::<(u32, u32)>::new();
            if i != 0 && grid[i - 1][j]  < v + 2 {
                new_edges.push((idx, to_idx(i - 1, j, cols) as u32));
            }
            if i != (grid.len() - 1) && grid[i + 1][j] < v + 2 {
                new_edges.push((idx, to_idx(i + 1, j, cols) as u32));
            }
            if j != 0 && grid[i][j - 1] < v + 2 {
                new_edges.push((idx, to_idx(i, j - 1, cols) as u32));
            }
            if j != (cols - 1) && grid[i][j + 1] < v + 2 {
                new_edges.push((idx, to_idx(i, j + 1, cols) as u32));
            }
            g.extend_with_edges(new_edges);
        }
    }
    g
}

fn to_node_index(node: &Node, cols: usize) -> NodeIndex {
    (to_idx(node.i, node.j, cols) as u32).try_into().unwrap()
}

fn problem_1(g: &Graph<u32, u32>, cols: usize, start: &Node, end: &Node)
    -> u32 {
    let start_node = to_node_index(start, cols);
    let end_node = to_node_index(end, cols);
    let res = dijkstra(g, start_node, Some(end_node), |_| 1);
    res[&end_node]
}

fn problem_2(grid: &Vec<Vec<u32>>,
             g: &mut Graph<u32, u32>,
             cols: usize,
             end: &Node) -> u32 {
    // reverse the direction of all the edges of the graph and set the "end"
    // node as the starting node for dijkstra
    g.reverse();
    let end_node = to_node_index(&end, cols);
    let res = dijkstra(&*g, end_node, None, |_| 1);
    // find the node with value 'a' (==0) with the minimum distance from "end"
    let mut min = u32::MAX;
    for i in 0..grid.len() {
        for j in 0..cols {
            let node = &to_node_index(&Node { i: i, j: j }, cols);
            // Note that the dijkstra result may not contain a node index if the
            // corresponding node is unreachable
            if res.contains_key(node) && grid[i][j] == 0 {
                min = cmp::min(min, res[node]);
            }
        }
    }
    min
}

fn main() {
    let filename = parse_args();
    let (grid, start, end, cols) = parse_input(&filename);
    let mut g = to_graph(&grid, cols);
    println!("Day 12, problem 1: {}", problem_1(&g, cols, &start, &end));
    println!("Day 12, problem 2: {}", problem_2(&grid, &mut g, cols, &end));
}
